src/route/dataset-route.ts

```diff
@@ -0,0 +1,176 @@
+import { Request, Response, Router } from 'express';
+import multer from 'multer';
+import pino from 'pino';
+
+// eslint-disable-next-line import/no-cycle
+import { processCSV, uploadCSV, DEFAULT_PAGE_SIZE } from '../controllers/csv-processor';
+import { DataLakeService } from '../controllers/datalake';
+import { Dataset } from '../entity/dataset';
+import { Datafile } from '../entity/datafile';
+import { FileDescription } from '../models/filelist';
+import { datasetToDatasetDTO } from '../dtos/dataset-dto';
+
+const storage = multer.memoryStorage();
+const upload = multer({ storage });
+
+export const logger = pino({
+    name: 'StatsWales-Alpha-App',
+    level: 'debug'
+});
+
+export const apiRoute = Router();
+
+apiRoute.post('/', upload.single('csv'), async (req: Request, res: Response) => {
+    if (!req.file) {
+        res.status(400);
+        res.json({
+            success: false,
+            headers: undefined,
+            data: undefined,
+            errors: [
+                {
+                    field: 'csv',
+                    message: 'No CSV data available'
+                }
+            ]
+        });
+        return;
+    }
+    const internalName: string = req.body?.internal_name;
+    if (!internalName) {
+        res.status(400);
+        res.json({
+            success: false,
+            errors: [
+                {
+                    field: 'internal_name',
+                    message: 'No internal name for the dataset has been provided'
+                }
+            ]
+        });
+        return;
+    }
+    const dataset = Dataset.createDataset(internalName, 'BetaUser');
+    const saved_dataset_record = await dataset.save();
+    const uploadDTO = await uploadCSV(req.file?.buffer, saved_dataset_record);
+    if (!uploadDTO.success) {
+        res.status(400);
+    }
+    res.json(uploadDTO);
+});
+
+apiRoute.get('/', async (req, res) => {
+    const datasets = await Dataset.find();
+    const fileList: FileDescription[] = [];
+    for (const dataset of datasets) {
+        fileList.push({
+            internal_name: dataset.internalName,
+            id: dataset.id
+        });
+    }
+    res.json({ filelist: fileList });
+});
+
+apiRoute.get('/:dataset', async (req, res) => {
+    const datasetID = req.params.dataset;
+    if (datasetID === undefined || datasetID === null) {
+        res.status(404);
+        res.json({ message: 'Dataset not found... file is null or undefined' });
+        return;
+    }
+    const dataset = await Dataset.findOneBy({ id: datasetID });
+    if (!dataset) {
+        res.status(404);
+        res.json({ message: 'Dataset not found... dataset id is invalid' });
+        return;
+    }
+    const datafiles = await dataset.datafiles;
+    if (datafiles.length < 1) {
+        res.status(404);
+        res.json({ message: 'Dataset has no datafiles attached' });
```

Comment:
Because that situation shouldn't happen currently...  It's not impossible... e.g we go in and delete a couple of datafiles from the database for example which results in us having a dataset with no datafiles which is why I handle the situation to say... 404 not found.  The next bits of the API route are about getting views of the fact table.  So we could make it a 500 (Internal server error), 410 (gone), 418 (I'm a teapot), etc.

As I evolve the backend based on in put from yourself and the user testing this of course could change quite rapidly.